src/planet-a/SubdivTriangleMesh.h

```diff
@@ -0,0 +1,487 @@
+            ChunkId chunkId, uint16_t x, uint16_t y) const noexcept
+    {
+        if (auto const [localId, shared] = coord_to_shared(x, y, m_chunkWidth);
+            shared)
+        {
+            return shared_get_vrtx(chunk_shared(chunkId)[size_t(localId)]);
+        }
+
+        // Center, non-shared chunk vertex
+        return VertexId(m_chunkVrtxFillCount * size_t(chunkId) + xy_to_triangular(x - 1, y - 2));
+    }
+
+    /**
+     * @return Number of triangles along the edge of a chunk
+     */
+    constexpr uint32_t chunk_width() const noexcept { return m_chunkWidth; }
+
+    /**
+     * @return Number of fill vertices per chunk
+     */
+    constexpr uint32_t chunk_vrtx_fill_count() const noexcept { return m_chunkVrtxFillCount; }
+
+    /**
+     * @brief Get shared vertices used by a specific chunk
+     *
+     * @param chunkId [in] Chunk Id
+     *
+     * @return Array view of shared vertices, access using ChunkLocalSharedId
+     */
+    ArrayView_t<SharedVrtxStorage_t const> chunk_shared(ChunkId chunkId) const noexcept
+    {
+        size_t const offset = size_t(chunkId) * m_chunkVrtxSharedCount;
+        return {&m_chunkSharedUsed[offset], m_chunkVrtxSharedCount};
+    };
+
+    /**
+     *
+     * Param 0: Newly added shared vertices; iterate this.
+     * Param 1: Maps SharedVrtxId to their associated SkVrtxId.
+     * Param 2: The index of the first shared vertex in the vertex buffer.
+     */
+    template<typename FUNC_T>
+    void shared_update(FUNC_T func)
+    {
+        func(ArrayView_t<SharedVrtxId const>(m_sharedNewlyAdded),
+             ArrayView_t<SkVrtxStorage_t const>(m_sharedSkVrtx.get(), m_sharedMax));
+
+        m_sharedNewlyAdded.clear();
+    }
+
+    /**
+     * @return Max number of shared vertices
+     */
+    constexpr uint32_t shared_count_max() const noexcept { return m_sharedMax; }
+
+    uint8_t& shared_face_count(SharedVrtxId sharedId) const noexcept { return m_sharedFaceCount[size_t(sharedId)]; }
+
+    /**
+     * @return Max number of mesh triangles / Required index buffer size.
+     */
+    constexpr uint32_t index_count_max() const noexcept { return m_indxCountMax; };
+
+    constexpr uint32_t index_chunk_offset(ChunkId chunkId) const noexcept { return m_chunkIndxFillCount * uint32_t(chunkId); }
+
+    /**
+     * @return Total max number of shared and fill vertices / Required vertex
+     *         buffer size.
+     */
+    constexpr uint32_t vertex_count_max() const noexcept { return m_vrtxCountMax; };
+
+    /**
+     * @return Index of first shared vertex / Max total number of fill vertices
+     */
+    constexpr uint32_t vertex_offset_shared() const noexcept { return m_vrtxSharedOffset; }
+
+    constexpr SharedVrtxId vertex_to_shared(VertexId sharedId) const noexcept { return SharedVrtxId(uint32_t(sharedId) - m_vrtxSharedOffset); }
+
+    constexpr uint32_t vertex_offset_fill(ChunkId chunkId) const noexcept
+    {
+        return m_chunkVrtxFillCount * uint32_t(chunkId);
```

Comment:
for types that are known to be strong typedefs like this, personally i recommend using std::to_underlying.

Reply:
Imo, that just sounds longer to write. In many places, I simply typedefed the underlying type.

In this case it needs to be a uint32, since ChunkId and m_chunkVrtxFillCount are both uint16. Just found out, that two uint16_t will multiply to an int.